ev3dev2/motor.py

```diff
@@ -2099,6 +2123,178 @@ def follow_line(self,
+            self.__gyro.mode = 'GYRO-RATE'
+            self.__gyro.mode = 'GYRO-ANG'
+            time.sleep(0.5)
+
+    def follow_gyro_angle(self,
+            kp, ki, kd,
+            speed,
+            target_angle=0,
+            sleep_time=0.01,
+            follow_for=follow_for_forever,
+            **kwargs
+        ):
+        """
+        PID gyro angle follower
+
+        ``kp``, ``ki``, and ``kd`` are the PID constants.
+
+        ``speed`` is the desired speed of the midpoint of the robot
+
+        ``target_angle`` is the angle we want to maintain
+
+        ``sleep_time`` is how many seconds we sleep on each pass through
+            the loop.  This is to give the robot a chance to react
+            to the new motor settings. This should be something small such
+            as 0.01 (10ms).
+
+        ``follow_for`` is called to determine if we should keep following the
+            desired angle or stop.  This function will be passed ``self`` (the current
+            ``MoveWithGyroTank`` object). Current supported options are:
+            - ``follow_for_forever``
+            - ``follow_for_ms``
+
+        ``**kwargs`` will be passed to the ``follow_for`` function
+
+        Example:
+
+        .. code:: python
+
+            from ev3dev2.motor import OUTPUT_A, OUTPUT_B, MoveTank, SpeedPercent, follow_for_ms
+            from ev3dev2.sensor.lego import GyroSensor
+
+            # Instantiate the MoveTank object
+            tank = MoveTank(OUTPUT_A, OUTPUT_B)
+
+            # Initialize the tank's gyro sensor
+            tank.gyro = GyroSensor()
+
+            try:
+                # Calibrate the gyro to eliminate drift, and to initialize the current angle as 0
+                tank.calibrate_gyro()
+
+                # Follow the line for 4500ms
+                tank.follow_gyro_angle(
+                    kp=11.3, ki=0.05, kd=3.2,
+                    speed=SpeedPercent(30),
+                    target_angle=0
+                    follow_for=follow_for_ms,
+                    ms=4500
+                )
+            except FollowGyroAngleErrorTooFast:
+                tank.stop()
+                raise
+        """
+        assert self.__gyro, "GyroSensor must be defined"
+
+        integral = 0.0
+        last_error = 0.0
+        derivative = 0.0
+        speed_native_units = speed.to_native_units(self.left_motor)
+        MAX_SPEED = SpeedNativeUnits(self.max_speed)
+
+        while follow_for(self, **kwargs):
+            current_angle = self.__gyro.angle
+            error = current_angle - target_angle
+            integral = integral + error
+            derivative = error - last_error
+            last_error = error
+            turn_native_units = (kp * error) + (ki * integral) + (kd * derivative)
+
+            left_speed = SpeedNativeUnits(speed_native_units - turn_native_units)
+            right_speed = SpeedNativeUnits(speed_native_units + turn_native_units)
+
+            if left_speed > MAX_SPEED:
```

Comment:
I think this needs to be `if abs(left_speed)...` in case the motor is being asked to move backwards at an impossible speed.  That seems very unlikely but you never know.

Same for the right_speed check on 2224